### ISSUE TITLE
Rethink ApolloError

### ISSUE THREAD

**Issue by glasser** [apollographql/apollo-server#6053]
ApolloError in apollo-server-errors adds some functionality to GraphQLError from graphql-js. graphql-js may have already caught up with this to some degree, and other functionality perhaps is general-purpose enough that it could easily just move to graphql-js. Let's figure out exactly what the value of having our own subclass is and see what we can do to reduce complexity.

**Comment by benjamn** [comment]
Another implementation of `ApolloError` is exported by `@apollo/client/errors` [here](https://github.com/apollographql/apollo-client/blob/main/src/errors/index.ts). I wonder if we can consolidate these implementations?

**Comment by glasser** [comment]
On `version-4` I moved the code from apollo-server-errors into the main package.

That said, it's possible we might decide that errors do deserve their own package, so that a *library* could implement an ApolloError (if it still exists) or throw a specific error like BadInputError without needing a dependency on `@apollo/server` that could pin a too-specific version. (eg, `@apollo/gateway` depends on `apollo-server-errors` and we'd really like it to not depend on `@apollo/server`.)

**Comment by glasser** [comment]
Worth looking at https://github.com/apollographql/apollo-server/pull/6316 to see if some of its ideas are relevant.

**Comment by glasser** [comment]
Major progress on version-4 in #6355. We still need to describe those changes and maybe make a few more like renaming formatApolloErrors.

**Comment by glasser** [comment]
In addition to writing the docs, @trevor-scheer and I have some questions we're thinking about with regards to the exported XError types. It's now quite easy to define your own subclasses with particular error codes. So the questions are:

- Do we really even need AuthenticationError and ForbiddenError, which are not used internally and are just provided for users?
- Do we actually need to export the other ones like SyntaxError, or should we just use it internally and document that people who want to see if an error is of a particular type should check the code extension instead?
- If we do want to export all of them then we should be consistent and export PersistedQueryNotSupportedError, OperationResolutionError, and BadRequestError.
- Would it be cool to have a little function like `const AuthenticationError = makeErrorClass('AuthenticationError', 'UNAUTHENTICATED')`? Or maybe not.

**Comment by IvanGoncharov** [comment]
> * Do we really even need AuthenticationError and ForbiddenError, which are not used internally and are just provided for users?

@glasser I agree we should remove them.

> * Do we actually need to export the other ones like SyntaxError, or should we just use it internally and document that people who want to see if an error is of a particular type should check the code extension instead?

No, I think exposing enum with codes would be enough.
So instead of `error instanceof SyntaxError` you can do `error.extensions?.code === ApolloServerErrorCode.GRAPHQL_PARSE_FAILED`.

> * If we do want to export all of them then we should be consistent and export PersistedQueryNotSupportedError, OperationResolutionError, and BadRequestError.

No, I think they don't add a lot of value.

> * Would it be cool to have a little function like `const AuthenticationError = makeErrorClass('AuthenticationError', 'UNAUTHENTICATED')`? Or maybe not.

This function feels more like ES5 (before modern ES6 class syntax).
It saves a few repetitive line:
```js
export class PersistedQueryNotSupportedError extends GraphQLErrorWithCode {
  constructor() {
    super(
      'PersistedQueryNotSupported',
      ApolloServerErrorCode.PERSISTED_QUERY_NOT_SUPPORTED,
    );
    this.name = 'PersistedQueryNotSupportedError';
  }
}
```
But I don't think that's a big saving compared to messing with class prototypes dynamically.

I implement above suggestions in #6705

**Comment by glasser** [comment]
With #6705 merged and #6696 incorporated into #6687 I think we can declare this done!
